crates/typst/src/export/svg.rs

```diff
@@ -0,0 +1,460 @@
+use std::{
+    collections::HashMap,
+    fmt::{Display, Write},
+    io::Read,
+};
+
+use base64::Engine;
+use ttf_parser::{GlyphId, OutlineBuilder};
+use usvg::{NodeExt, TreeParsing};
+
+use crate::{
+    doc::{Document, Frame, FrameItem, Glyph, GroupItem, TextItem},
+    geom::{Abs, Axes, Geometry, LineCap, LineJoin, Shape, Transform},
+    image::{ImageFormat, RasterFormat, VectorFormat},
+    util::hash128,
+};
+use crate::{geom::Paint::Solid, image::Image};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+struct RenderHash(u128);
+
+impl From<u128> for RenderHash {
+    fn from(value: u128) -> Self {
+        Self(value)
+    }
+}
+
+impl Display for RenderHash {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        base64::engine::general_purpose::STANDARD
+            .encode(self.0.to_le_bytes())
+            .fmt(f)
+    }
+}
+
+/// Export a document into a SVG file.
+pub fn svg(doc: &Document) -> String {
+    let mut renderer = SVGRenderer::default();
+    let mut max_width = Abs::zero();
+    let mut y_offset = Abs::zero();
+    for page in &doc.pages {
+        let page_string =
+            renderer.render_frame(page, Transform::translate(Abs::zero(), y_offset));
+        renderer.append_page(page_string);
+        y_offset += page.size().y;
+        max_width = max_width.max(page.size().x);
+    }
+    let doc_size = Axes { x: max_width, y: y_offset };
+    renderer.finalize(doc_size)
+}
+
+#[derive(Debug, Clone, Default)]
+struct SVGRenderer {
+    body: String,
+    glyphs: HashMap<RenderHash, String>,
+    clip_paths: HashMap<RenderHash, String>,
```

Comment:
How do you think of the hash used here? Should we put the more info about the glyph into the key of the hashmap to avoid possible hash collision?

Reply:
I think I'd prefer to just use ascending ids instead of hashes so that we get the same output even if built with a newer compiler (that might generate different hashes). And we don't need to think about collisions at all. Also, since `HashMap` iteration order is random, right now the order of things in the output might change even from execution to execution. It would be better to use something ordered here. You could store things in a vector (with the index being the unique id) and only perform hash-based deduplication through a hashmap. Similar things are happening in the PDF export I believe.